_posts/drafts/go-object-model.md

```diff
@@ -0,0 +1,286 @@
+---
+published: false
+title: Fun with Go's object model
+layout: post
+---
+Go object model gets easier to understand when you accept
+that there is no objects at all, there is just functions and
+contracts, that are basically sets of functions.
+
+<!-- more -->
+
+Ok, now that I got your attention and you must be thinking
+"shut up, of course there are objects in Go", well, you are
+right. But it is a object model that is very different from
+the classical ones (like Java, C++, Python, the ones I know
+at least).
+
+Since I had my share of C programming, and also other cool
+languages like Lua, it is not very hard to me to think
+about closures and functions, but as I see other people
+developing in Go I sense some struggling to understand how
+Go works with objects.
+
+A vivid example is not seeing that a function that accepts
+a function as a parameter can also be called passing a method,
+So the idea is to de construct the object model to just functions
+and build it back to how Go works, showing that the first class
+concept in Go is actually functions not objects.
+
+It will probably be a lousy de construction since I'm not an
+expert in Go, but yet I feel compelled to try :-).
+
+
+# In the beginning there was the functions
+
+
+Well I'm trying to make a point around thinking first about
+functions, so here goes an lousy example of what can be done with functions.
+
+Lets define a type that is actually a function:
+
+```go
+type Adder func(int, int) int
+```
+
+You can think about it as a interface (but it is not the same thing).
+Any function that matches the signature will be acceptable as being of
```

Comment:
A good way to explain is decomposing the function name from the function declaration. Go func declaration, as in many other languages, could be thought as only a syntax sugar also, to something like:

```go
var add = func(x, y int) { return x+y }
```
Then, only the value of the assignment matter to match the type Adder.